Samples/ObservableScreen/Sources/MultiCounterWorkflow.swift

```diff
@@ -0,0 +1,125 @@
+        var max: CounterInfo = .init(name: "Max")
+        var nextCounter = 1
+
+        var resetToken = CounterWorkflow.ResetToken()
+
+        mutating func addCounter() {
+            counters += [.init(name: "Counter \(nextCounter)")]
+            nextCounter += 1
+        }
+    }
+
+    func makeInitialState() -> State {
+        var state = State()
+        state.addCounter()
+        state.addCounter()
+        return state
+    }
+
+    struct ResetAction: WorkflowAction {
+        typealias WorkflowType = MultiCounterWorkflow
+
+        func apply(toState state: inout MultiCounterWorkflow.State) -> Never? {
+            state.resetToken = .init()
+            return nil
+        }
+    }
+
+    @CasePathable
+    enum SumAction: WorkflowAction {
+        typealias WorkflowType = MultiCounterWorkflow
+
+        case showSum(Bool)
+
+        func apply(toState state: inout MultiCounterWorkflow.State) -> Never? {
+            switch self {
+            case .showSum(let showSum):
+                state.showSum = showSum
+                return nil
+            }
+        }
+    }
+
+    enum CounterAction: WorkflowAction {
+        typealias WorkflowType = MultiCounterWorkflow
+
+        case addCounter
+        case removeCounter(UUID)
+
+        func apply(toState state: inout MultiCounterWorkflow.State) -> Never? {
+            switch self {
+            case .addCounter:
+                state.addCounter()
+                return nil
+            case .removeCounter(let id):
+                state.counters.removeAll { $0.id == id }
+                return nil
+            }
+        }
+    }
+
+    typealias Output = Never
+    typealias Rendering = MultiCounterModel
+
+    func render(state: State, context: RenderContext<MultiCounterWorkflow>) -> Rendering {
+        print("\(Self.self) rendered")
+
+        let maxCounter: CounterWorkflow.Model? = if state.showMax {
+            CounterWorkflow(
+                info: state.max,
+                resetToken: state.resetToken,
+                initialValue: 5
+            )
+            .rendered(in: context, key: "max")
+        } else {
+            nil
+        }
+
+        let counters: [CounterWorkflow.Model] = state.counters.map { counter in
+            CounterWorkflow(
+                info: counter,
+                resetToken: state.resetToken,
+                maxValue: maxCounter?.count
+            )
+            .rendered(in: context, key: "\(counter.id)")
+        }
+
+        let sumAction = context.makeSink(of: SumAction.self)
+        let counterAction = context.makeSink(of: CounterAction.self)
+        let resetAction = context.makeSink(of: ResetAction.self)
+
+        return MultiCounterModel(
+            accessor: context.makeStateAccessor(state: state),
+            counters: counters,
+            maxCounter: maxCounter,
+            sumAction: sumAction,
+            counterAction: counterAction,
+            resetAction: resetAction
+        )
+    }
```

Comment:
Something that's confusing to me is what this is rendering and how it relates to the views / screen. Sharing the thought process I went through when trying to review this.

It looks like the new expectation is Workflows render a model instead of a screen. But how does that model relate to the eventual screen / view that gets displayed? My first guess was to look at `MultiCounterModel` and see if it has something that produces a screen / view, but it doesn't, so I hit a bit of a dead end because the Workflow doesn't seem to expose any association to the view that gets displayed.

So then I searched around for `MultiCounterView` and found the only place that creates it is the `MultiCounterScreen`. Then I searched for who creates that, and it's actually created by the code that uses `MultiCounterWorkflow`:

```swift
let root = WorkflowHostingController(
    workflow: MultiCounterWorkflow().mapRendering(MultiCounterScreen.init)
)
```

This is jarring because it means that an individual workflow type (in this case `MultiCounterWorkflow`) has no direct tie to the actual screen / view that uses the data from the Workflow. Perhaps this is the intention, but it feels like a weird use of Workflow's paradigms. For example, the name of "rendering" no longer seems to fit here - Workflows don't produce a rendering anymore, they produce a model that an unrelated screen knows how to convert to a SwiftUI view. And if I'm a developer trying to consume a Workflow from another module, how do I know what view I'm supposed to map the rendering to? could we tie them together a bit more so developers don't have to literally read the code to figure out what model connects to which screen?

Reply:
Not sure if this would work at all but something that comes to mind is composing the two concepts together, so you have:
* A workflow that renders a model, this implementation of `MultiCounterWorkflow`
* A workflow that renders a screen / view, and consumes the other workflow (in this case `MultiCounterWorkflow`) as a child in it's `render` function and does the translation from `MultiCounterWorkflow.Rendering` to `MultiCounterScreen`

Reply:
Good feedback! I'll break it up a bit to respond.

> This is jarring because it means that an individual workflow type (in this case `MultiCounterWorkflow`) has no direct tie to the actual screen / view that uses the data from the Workflow. Perhaps this is the intention, but it feels like a weird use of Workflow's paradigms. 

That's correct. The view and the workflow both have a dependency on the state and actions (collectively, the model), but not necessarily on each other. If the state & actions are nested in the workflow type, then the view has a dependency on that, effectively inverting the dependency you'd see if rendering a screen. It's a different pattern, for sure, and it's something we expected to require some education. I'm working on an adoption guide to help with that.

> For example, the name of "rendering" no longer seems to fit here - Workflows don't produce a rendering anymore, they produce a model that an unrelated screen knows how to convert to a SwiftUI view. 

The rendering is, by definition, whatever the workflow renders. The notion that every workflow should be rendering a screen is so pervasive today that I'd say it's harmful — nothing is composable and tests have to force-unwrap type erasers. In a world where views are easily composable, you often won't need a screen to act as the glue, so I'm trying to discourage the idea that every workflow has associated screen type.

> And if I'm a developer trying to consume a Workflow from another module, how do I know what view I'm supposed to map the rendering to? could we tie them together a bit more so developers don't have to literally read the code to figure out what model connects to which screen?

Well in most cases I think it'll be obvious, because they'll live in the same module and have a naming pattern. In other cases, it's a matter of matching up the model. There's no guaranteed 1:1 mapping — you could write 2 different view implementations that take the same model type, for example. You could even power a view with 2 different workflows, but I think that's unlikely.

> Not sure if this would work at all but something that comes to mind is composing the two concepts together, so you have:
> 
> * A workflow that renders a model, this implementation of `MultiCounterWorkflow`
> * A workflow that renders a screen / view, and consumes the other workflow (in this case `MultiCounterWorkflow`) as a child in it's `render` function and does the translation from `MultiCounterWorkflow.Rendering` to `MultiCounterScreen`

You could write a workflow that does that latter step, but that's exactly what `mapRendering()` does!

Reply:
> You could write a workflow that does that latter step, but that's exactly what mapRendering() does!

Fair! But I was looking for a way for consumers to not have to dig into the code to know what to map the rendering to. Sure they might be similarly named but its still an awkward interface.

> The rendering is, by definition, whatever the workflow renders. The notion that every workflow should be rendering a screen is so pervasive today that I'd say it's harmful — nothing is composable and tests have to force-unwrap type erasers. In a world where views are easily composable, you often won't need a screen to act as the glue, so I'm trying to discourage the idea that every workflow has associated screen type.

100% understand this - I'm more thinking of it as a developer coming the existing norm to this.

I guess the overall feeling is that this doesn't really fit into how we use Workflow today but its still built on top of Workflow as we know it. I wonder if theres some way we could separate the two a bit. What would it look like if you were to implement this from scratch, and how would you want to integrate that into an existing codebase that uses Workflow?

Reply:
> I guess the overall feeling is that this doesn't really fit into how we use Workflow today but its still built on top of Workflow as we know it. I wonder if theres some way we could separate the two a bit. What would it look like if you were to implement this from scratch, and how would you want to integrate that into an existing codebase that uses Workflow?

Well, you're not wrong, but I think that was sort of the goal, to build a SwiftUI integration on top of Workflow. I'm not sure how we could separate the two further without breaking that. We did prototype some alternative designs that were much more similar to how Workflow is used today, but ultimately opted for this instead because its advantages seemed to be worth the cost of the pattern disruption.